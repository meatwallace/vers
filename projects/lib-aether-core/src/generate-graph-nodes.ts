import invariant from 'tiny-invariant';
import { createAetherNode } from './create-aether-node';
import { AetherNode } from './types';

/**
 * Generates a graph of AetherNodes and returns them as an array.
 *
 * @param maxDifficulty - The maximum difficulty of the graph
 * @returns A graph of AetherNodes
 */
export function generateGraphNodes(maxDifficulty: number): Array<AetherNode> {
  const graph: Array<AetherNode> = [];

  const centralNode = createAetherNode(0, 0);

  graph.push(centralNode);

  // generate all the nodes in our graph
  for (let difficulty = 1; difficulty <= maxDifficulty; difficulty++) {
    // linear growth in nodes per difficulty
    const nodesInLevel = 4 * difficulty;

    for (let i = 0; i < nodesInLevel; i++) {
      graph.push(createAetherNode(i, difficulty));
    }
  }

  // set all our connections
  for (const [i, node] of graph.entries()) {
    invariant(node, 'node is required');

    const nodesInCurrentLevel = 4 * node.difficulty;
    const nodesInPreviousLevel = nodesInCurrentLevel - 4;
    const nodesInNextLevel = nodesInCurrentLevel + 4;

    const startIndexOfCurrentLevel = i - node.index;
    const startIndexOfPreviousLevel =
      startIndexOfCurrentLevel - nodesInPreviousLevel;
    const startIndexOfNextLevel =
      startIndexOfCurrentLevel + nodesInCurrentLevel;

    // we'll manually attach our origin node at the end
    if (node.difficulty > 0) {
      let prevIndex1: number;
      let prevIndex2: number;
      let nextIndex1: number;
      let nextIndex2: number;

      const segment = Math.floor(node.index / node.difficulty);

      // a good mental model for this is as we get further away from our +X axis (which is represented
      // by our segment), the connections to the next highest difficulty increase in distance, accounting
      // for the linear growth.
      switch (segment) {
        case 0: {
          prevIndex1 = i - nodesInPreviousLevel;
          prevIndex2 = prevIndex1 - 1;
          nextIndex1 = i + nodesInCurrentLevel;
          nextIndex2 = nextIndex1 + 1;

          // if it is the first node of the difficulty, we need to connect to the last node of the next
          if (node.index / node.difficulty === 0) {
            prevIndex2 = startIndexOfNextLevel + nodesInNextLevel - 1;
          }

          break;
        }

        case 1: {
          prevIndex1 = i - nodesInPreviousLevel - 1;
          prevIndex2 = prevIndex1 - 1;
          nextIndex1 = i + nodesInCurrentLevel + 1;
          nextIndex2 = nextIndex1 + 1;

          // we're travelling along our axis, so connect to a the higher difficulty node
          if (node.index / node.difficulty === 1) {
            prevIndex2 = startIndexOfNextLevel + node.index;
          }

          break;
        }

        case 2: {
          prevIndex1 = i - nodesInPreviousLevel - 2;
          prevIndex2 = prevIndex1 - 1;
          nextIndex1 = i + nodesInCurrentLevel + 2;
          nextIndex2 = nextIndex1 + 1;

          // we're travelling along our axis, so connect to a the higher difficulty node
          if (node.index / node.difficulty === 2) {
            prevIndex2 = startIndexOfNextLevel + node.index + 1;
          }

          break;
        }

        case 3: {
          prevIndex1 = i - nodesInPreviousLevel - 3;
          prevIndex2 = prevIndex1 - 1;
          nextIndex1 = i + nodesInCurrentLevel + 3;
          nextIndex2 = nextIndex1 + 1;

          // we're travelling along our axis, so connect to a the higher difficulty node
          if (node.index / node.difficulty === 3) {
            prevIndex2 = startIndexOfNextLevel + node.index + 2;
          }

          // if its the last node of the difficulty, we need to connect to the first node of the previous
          if (node.index + 1 === nodesInCurrentLevel) {
            prevIndex1 = startIndexOfPreviousLevel;
          }

          break;
        }

        default: {
          invariant(false, 'invalid segment');
        }
      }

      const connection1 = graph[prevIndex1];
      const connection2 = graph[prevIndex2];
      const connection3 = graph[nextIndex1];
      const connection4 = graph[nextIndex2];

      node.connections = [
        connection1?.id ?? null,
        connection2?.id ?? null,
        connection3?.id ?? null,
        connection4?.id ?? null,
      ];
    }
  }

  // connect the central node to the first difficulty
  for (let i = 0; i < 4; i++) {
    const node = graph[i + 1];

    invariant(node, 'node is required');

    centralNode.connections[i] = node.id;
    node.connections[0] = centralNode.id;
  }

  return graph;
}
