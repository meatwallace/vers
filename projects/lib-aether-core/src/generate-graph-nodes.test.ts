import { expect, test, vi } from 'vitest';
import { generateGraphNodes } from './generate-graph-nodes';

// mock our ID & seed generation so they return predictable values
vi.mock('./create-id', () => {
  let id = 0;

  const createID = vi.fn(() => {
    const result = `id-${id}`;
    id++;
    return result;
  });

  return { createID };
});

vi.mock('./get-seed', () => {
  let seed = 0;

  const getSeed = vi.fn(() => seed++);

  return { getSeed };
});

// rather than testing our implementation, snapshot a valid graph.
// if this changes we need to go over it with a fine tooth comb.
test('it generates a valid graph', () => {
  const nodes = generateGraphNodes(3);

  expect(nodes).toMatchSnapshot();
});

test('it generates a central origin node', () => {
  const nodes = generateGraphNodes(1);

  expect(nodes[0]).toStrictEqual({
    connections: expect.arrayContaining([
      expect.any(String),
      expect.any(String),
      expect.any(String),
      expect.any(String),
    ]),
    difficulty: 0,
    id: expect.any(String),
    index: 0,
    position: [0, 0],
    seed: expect.any(Number),
  });
});

test('it generates the correct number of nodes for each difficulty level', () => {
  const nodes = generateGraphNodes(3);

  // 1 + 4 + 8 + 12 = 25
  expect(nodes).toHaveLength(25);

  const difficulty1Nodes = nodes.filter((node) => node.difficulty === 1);
  const difficulty2Nodes = nodes.filter((node) => node.difficulty === 2);
  const difficulty3Nodes = nodes.filter((node) => node.difficulty === 3);

  expect(difficulty1Nodes).toHaveLength(4);
  expect(difficulty2Nodes).toHaveLength(8);
  expect(difficulty3Nodes).toHaveLength(12);
});
