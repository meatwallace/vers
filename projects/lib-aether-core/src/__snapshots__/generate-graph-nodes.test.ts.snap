// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`it generates a valid graph 1`] = `
[
  {
    "connections": [
      "id-1",
      "id-2",
      "id-3",
      "id-4",
    ],
    "difficulty": 0,
    "id": "id-0",
    "index": 0,
    "position": [
      0,
      0,
    ],
    "seed": 0,
  },
  {
    "connections": [
      "id-0",
      "id-12",
      "id-5",
      "id-6",
    ],
    "difficulty": 1,
    "id": "id-1",
    "index": 0,
    "position": [
      1,
      0,
    ],
    "seed": 1,
  },
  {
    "connections": [
      "id-0",
      "id-6",
      "id-7",
      "id-8",
    ],
    "difficulty": 1,
    "id": "id-2",
    "index": 1,
    "position": [
      0,
      1,
    ],
    "seed": 2,
  },
  {
    "connections": [
      "id-0",
      "id-8",
      "id-9",
      "id-10",
    ],
    "difficulty": 1,
    "id": "id-3",
    "index": 2,
    "position": [
      -1,
      0,
    ],
    "seed": 3,
  },
  {
    "connections": [
      "id-0",
      "id-10",
      "id-11",
      "id-12",
    ],
    "difficulty": 1,
    "id": "id-4",
    "index": 3,
    "position": [
      0,
      -1,
    ],
    "seed": 4,
  },
  {
    "connections": [
      "id-1",
      "id-24",
      "id-13",
      "id-14",
    ],
    "difficulty": 2,
    "id": "id-5",
    "index": 0,
    "position": [
      2,
      0,
    ],
    "seed": 5,
  },
  {
    "connections": [
      "id-2",
      "id-1",
      "id-14",
      "id-15",
    ],
    "difficulty": 2,
    "id": "id-6",
    "index": 1,
    "position": [
      1.414,
      1.414,
    ],
    "seed": 6,
  },
  {
    "connections": [
      "id-2",
      "id-15",
      "id-16",
      "id-17",
    ],
    "difficulty": 2,
    "id": "id-7",
    "index": 2,
    "position": [
      0,
      2,
    ],
    "seed": 7,
  },
  {
    "connections": [
      "id-3",
      "id-2",
      "id-17",
      "id-18",
    ],
    "difficulty": 2,
    "id": "id-8",
    "index": 3,
    "position": [
      -1.414,
      1.414,
    ],
    "seed": 8,
  },
  {
    "connections": [
      "id-3",
      "id-18",
      "id-19",
      "id-20",
    ],
    "difficulty": 2,
    "id": "id-9",
    "index": 4,
    "position": [
      -2,
      0,
    ],
    "seed": 9,
  },
  {
    "connections": [
      "id-4",
      "id-3",
      "id-20",
      "id-21",
    ],
    "difficulty": 2,
    "id": "id-10",
    "index": 5,
    "position": [
      -1.414,
      -1.414,
    ],
    "seed": 10,
  },
  {
    "connections": [
      "id-4",
      "id-21",
      "id-22",
      "id-23",
    ],
    "difficulty": 2,
    "id": "id-11",
    "index": 6,
    "position": [
      0,
      -2,
    ],
    "seed": 11,
  },
  {
    "connections": [
      "id-1",
      "id-4",
      "id-23",
      "id-24",
    ],
    "difficulty": 2,
    "id": "id-12",
    "index": 7,
    "position": [
      1.414,
      -1.414,
    ],
    "seed": 12,
  },
  {
    "connections": [
      "id-5",
      null,
      null,
      null,
    ],
    "difficulty": 3,
    "id": "id-13",
    "index": 0,
    "position": [
      3,
      0,
    ],
    "seed": 13,
  },
  {
    "connections": [
      "id-6",
      "id-5",
      null,
      null,
    ],
    "difficulty": 3,
    "id": "id-14",
    "index": 1,
    "position": [
      2.598,
      1.5,
    ],
    "seed": 14,
  },
  {
    "connections": [
      "id-7",
      "id-6",
      null,
      null,
    ],
    "difficulty": 3,
    "id": "id-15",
    "index": 2,
    "position": [
      1.5,
      2.598,
    ],
    "seed": 15,
  },
  {
    "connections": [
      "id-7",
      null,
      null,
      null,
    ],
    "difficulty": 3,
    "id": "id-16",
    "index": 3,
    "position": [
      0,
      3,
    ],
    "seed": 16,
  },
  {
    "connections": [
      "id-8",
      "id-7",
      null,
      null,
    ],
    "difficulty": 3,
    "id": "id-17",
    "index": 4,
    "position": [
      -1.5,
      2.598,
    ],
    "seed": 17,
  },
  {
    "connections": [
      "id-9",
      "id-8",
      null,
      null,
    ],
    "difficulty": 3,
    "id": "id-18",
    "index": 5,
    "position": [
      -2.598,
      1.5,
    ],
    "seed": 18,
  },
  {
    "connections": [
      "id-9",
      null,
      null,
      null,
    ],
    "difficulty": 3,
    "id": "id-19",
    "index": 6,
    "position": [
      -3,
      0,
    ],
    "seed": 19,
  },
  {
    "connections": [
      "id-10",
      "id-9",
      null,
      null,
    ],
    "difficulty": 3,
    "id": "id-20",
    "index": 7,
    "position": [
      -2.598,
      -1.5,
    ],
    "seed": 20,
  },
  {
    "connections": [
      "id-11",
      "id-10",
      null,
      null,
    ],
    "difficulty": 3,
    "id": "id-21",
    "index": 8,
    "position": [
      -1.5,
      -2.598,
    ],
    "seed": 21,
  },
  {
    "connections": [
      "id-11",
      null,
      null,
      null,
    ],
    "difficulty": 3,
    "id": "id-22",
    "index": 9,
    "position": [
      0,
      -3,
    ],
    "seed": 22,
  },
  {
    "connections": [
      "id-12",
      "id-11",
      null,
      null,
    ],
    "difficulty": 3,
    "id": "id-23",
    "index": 10,
    "position": [
      1.5,
      -2.598,
    ],
    "seed": 23,
  },
  {
    "connections": [
      "id-5",
      "id-12",
      null,
      null,
    ],
    "difficulty": 3,
    "id": "id-24",
    "index": 11,
    "position": [
      2.598,
      -1.5,
    ],
    "seed": 24,
  },
]
`;
